c/datasetjson.c

```diff
@@ -2220,133 +2229,377 @@ int decodePercentByte(char *inString, int inLength, char *outString, int *outStr
+      }
+      recordsWritten++;
+    } else if (ferror(inDataset)) {
+      respondWithError(response,HTTP_STATUS_INTERNAL_SERVER_ERROR,"Error writing to dataset");
+      zowelog(NULL, LOG_COMP_RESTDATASET, ZOWE_LOG_DEBUG,  "Error reading DSN=%s, rc=%d\n", sourceDataset, bytesRead);
+      return;
+    }
+  }
+
+  if (!rcEtag) { rcEtag = icsfDigestFinish(&digest, hash); }
+  if (rcEtag) {
+    zowelog(NULL, LOG_COMP_RESTDATASET, ZOWE_LOG_WARNING,  "ICSF error for SHA etag, %d\n",rcEtag);
+  }
+
+  jsonPrinter *p = respondWithJsonPrinter(response);
+  setResponseStatus(response, 201, "Successfully Copied Dataset");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(p);
+
+  char msgBuffer[128];
+  snprintf(msgBuffer, sizeof(msgBuffer), "Pasted dataset %s with %d records", targetDataset, recordsWritten);
+  jsonAddString(p, "msg", msgBuffer);
+
+  if (!rcEtag) {
+    // Convert hash text to hex.
+    int eTagLength = digest.hashLength*2;
+    char eTag[eTagLength+1];
+    memset(eTag, '\0', eTagLength);
+    int len = digest.hashLength;
+    simpleHexPrint(eTag, hash, digest.hashLength);
+    jsonAddString(p, "etag", eTag);
+  }
+
+  jsonEnd(p);
+
+  finishResponse(response);
+
+  fclose(inDataset);
+  fclose(outDataset);
+
+  return;
+}
+
+void readAndWriteToDataset(HttpResponse *response, char* sourceDataset, char* targetDataset) {
+
+  DatasetName dsn;
+  DatasetMemberName memberName;
+  extractDatasetAndMemberName(sourceDataset, &dsn, &memberName);
+
+  DynallocDatasetName daDsn;
+  DynallocMemberName daMember;
+  memcpy(daDsn.name, dsn.value, sizeof(daDsn.name));
+  memcpy(daMember.name, memberName.value, sizeof(daMember.name));
+  DynallocDDName daDDname = {.name = "????????"};
+
+  int daRC = RC_DYNALLOC_OK, daSysRC = 0, daSysRSN = 0;
+  daRC = dynallocAllocDataset(
+      &daDsn,
+      IS_DAMEMBER_EMPTY(daMember) ? NULL : &daMember,
+      &daDDname,
+      DYNALLOC_DISP_SHR,
+      DYNALLOC_ALLOC_FLAG_NO_CONVERSION | DYNALLOC_ALLOC_FLAG_NO_MOUNT,
+      &daSysRC, &daSysRSN
+  );
+
+  if (daRC != RC_DYNALLOC_OK) {
+    zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+      	    "error: ds alloc dsn=\'%44.44s\', member=\'%8.8s\', dd=\'%8.8s\',"
+            " rc=%d sysRC=%d, sysRSN=0x%08X (read)\n",
+            daDsn.name, daMember.name, daDDname.name, daRC, daSysRC, daSysRSN);
+    respondWithDYNALLOCError(response, daRC, daSysRC, daSysRSN,
+                             &daDsn, &daMember, "r");
+    return;
+  }
+  zowelog(NULL, LOG_COMP_DATASERVICE, ZOWE_LOG_DEBUG,
+          "debug: reading dsn=\'%44.44s\', member=\'%8.8s\', dd=\'%8.8s\'\n",
+          daDsn.name, daMember.name, daDDname.name);
+
+  DDName ddName;
+  memcpy(&ddName.value, &daDDname.name, sizeof(ddName.value));
+
+  char ddPath[16];
+  snprintf(ddPath, sizeof(ddPath), "DD:%8.8s", ddName.value);
+
+  int lrecl = getLreclOrRespondError(response, &dsn, ddPath);
+  if (!lrecl) {
+    return;
```

Comment:
i think a `dynallocUnallocDatasetByDDName` needs to occur with this return, else the server holds some sort of lock on the dataset?

we see a `dynallocUnallocDatasetByDDName` on line 2448 so just make sure there's one here too, i think.